src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -701,58 +611,108 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
+    }
 
-  def annotateSamplesKeyTable(kt: KeyTable, vdsKey: java.util.ArrayList[String], code: String): VariantSampleMatrix[T] =
-    annotateSamplesKeyTable(kt, vdsKey.asScala, code)
+    val (f, liftedSignature): (Annotation => Annotation, Type) = if (product)
+      kt.valueSignature.size match {
+        case 0 => ((x: Annotation) => if (x == null) 0 else x.asInstanceOf[IndexedSeq[_]].length, TInt)
+        case 1 => ((x: Annotation) =>
+          if (x == null)
+            x
+          else
+            x.asInstanceOf[IndexedSeq[_]].map(r => if (r != null) r.asInstanceOf[Row].get(0) else null),
+          TArray(kt.valueSignature.fields.head.typ))
+        case _ => (identity[Annotation], TArray(kt.valueSignature))
+      } else
+      kt.valueSignature.size match {
+        case 0 => ((x: Annotation) => x != null, TBoolean)
+        case 1 => ((x: Annotation) => if (x == null) null else x.asInstanceOf[Row].get(0),
+          kt.valueSignature.fields.head.typ)
+        case _ => (identity[Annotation], kt.valueSignature)
+      }
 
-  def annotateSamplesKeyTable(kt: KeyTable, vdsKey: Seq[String], code: String): VariantSampleMatrix[T] = {
-    val vdsKeyEc = EvalContext(Map("s" -> (0, TString), "sa" -> (1, saSignature)))
+    val (finalType, inserter): (Type, (Annotation, Annotation) => Annotation) = {
+      val (t, ins) = if (isExpr) {
+        val ec = EvalContext(Map(
+          "sa" -> (0, saSignature),
+          "table" -> (1, liftedSignature)))
+        Annotation.buildInserter(annotationExpr, saSignature, ec, Annotation.SAMPLE_HEAD)
+      } else insertSA(liftedSignature, Parser.parseAnnotationRoot(annotationExpr, Annotation.SAMPLE_HEAD))
 
-    val (vdsKeyType, vdsKeyFs) = vdsKey.map(Parser.parseExpr(_, vdsKeyEc)).unzip
+      (t, (a: Annotation, toIns: Annotation) => ins(a, f(toIns)))
+    }
 
     val keyTypes = kt.keyFields.map(_.typ)
-    if (!(keyTypes sameElements vdsKeyType))
-      fatal(s"Key signature of KeyTable, `${ keyTypes.mkString(", ") }', must match type of computed key, `${ vdsKeyType.mkString(", ") }'.")
 
-    val inserterEc = EvalContext(Map("sa" -> (0, saSignature), "table" -> (1, kt.valueSignature)))
+    val keyedRDD = kt.keyedRDD()
+      .filter { case (k, v) => k.toSeq.forall(_ != null) }
 
-    val (finalType, inserter) =
-      buildInserter(code, vaSignature, inserterEc, Annotation.SAMPLE_HEAD)
+    Option(vdsKey) match {
+      case Some(keys) =>
+        val keyEC = EvalContext(Map("s" -> (0, TString), "sa" -> (1, saSignature)))
+        val (vdsKeyType, vdsKeyFs) = keys.map(Parser.parseExpr(_, keyEC)).unzip
 
-    val ktRdd = kt.keyedRDD()
+        if (!(keyTypes sameElements vdsKeyType))
+          fatal(
+            s"""method `annotateSamplesTable' encountered a mismatch between table keys and computed keys.
+               |  Computed keys:  [ ${ vdsKeyType.mkString(", ") } ]
+               |  Key table keys: [ ${ keyTypes.mkString(", ") } ]""".stripMargin)
```

Comment:
It mirrors python list representation. The reason I put them in here is that either one of these could be empty. Seeing:
```
 Computed keys: [ ]
```
is a bit clearer than

```
 Computed keys: 
```